### PR TITLE
warning added to output header of yarp clean command

### PR DESCRIPTION
**Pull request by aerydna**
added an important warning on the initial output of yarp clean.

### CODE REVIEW

File: src/libYARP_OS/src/Companion.cpp [hunk 1]
```diff
@@ -1482,6 +1482,15 @@ int Companion::cmdNamespace(int argc, char *argv[]) {
 int Companion::cmdClean(int argc, char *argv[]) {
     Property options;
     if (argc==0) {
+        printf("\
+##### ATTENTION #######\n\n \
+the 'yarp clean'' command cannot solve any Address Conflict.\n \
```
Comment (drdanz) [inline src/libYARP_OS/src/Companion.cpp]: too many ' after clean, address conflict should be lower case

File: src/libYARP_OS/src/Companion.cpp [hunk 2]
```diff
@@ -1482,6 +1482,15 @@ int Companion::cmdNamespace(int argc, char *argv[]) {
 int Companion::cmdClean(int argc, char *argv[]) {
     Property options;
     if (argc==0) {
+        printf("\
+##### ATTENTION #######\n\n \
```
Comment (drdanz) [inline src/libYARP_OS/src/Companion.cpp]: This should probably be WARNING, not attention.

File: src/libYARP_OS/src/Companion.cpp [hunk 3]
```diff
@@ -1482,6 +1482,15 @@ int Companion::cmdNamespace(int argc, char *argv[]) {
...
+the 'yarp clean'' command cannot solve any Address Conflict.\n \
+if not used properly it can cause them instead, deregistering\n \
+active ports from the name server\n \
+when the network is particulary slow\n \
+or if the timeout is set too quick\n\n \
+##### ATTENTION #######\n\n");
```
Comment (drdanz) [inline src/libYARP_OS/src/Companion.cpp]: The second "ATTENTION" is not needed...

**Comment by barbalberto** [comment]
I understand the idea, but adding a warning like this is only polluting the output and creating panic in the user.
I'm definitely **against** this message. If yarp clean have issues, fix the `yarp clean` instead.

**Comment by drdanz** [comment]
I agree with @barbalberto

**Comment by traversaro** [comment]
As a compromise between the user-scaring big warning message and the need to educate the users, I suggest to just add line pointing to a nice documentation page explaining what `yarp clean` commands does and which problems can solve (or introduce), and which problems it cannot influence.

**Comment by traversaro** [comment]
The documentation page is already there, see http://www.yarp.it/yarp.html#yarp_clean . I think we can extend if it does not contain all the information necessary to properly educate the users.

**Comment by aerydna** [comment]
yarp clean is not intuitive, fixing it seems not the priority at the moment and seems time expensive (honestly i don't even know if it is possible or not.. the requirement will be to check port avaiability remotely). for a similar cases (rosnode cleanup) ROS  print a similar warning and i think is a good idea for this particular delicate case.

**Comment by randaz81** [comment]
> The documentation page is already there, see http://www.yarp.it/yarp.html#yarp_clean . I think we can extend if it does not contain all the information necessary to properly educate the users.

As it usually happens for QA stuff, the long explanation here is forgotten: https://github.com/robotology/QA/issues/181
This open a different issue for me, which is: QA is not the right place to put this kind of tutorials (I'm ready to bet that 0% of students have read it, looking at how the yarp clean command is used indiscriminately). Please help me to find the right place to move this stuff to (It can even be the web page aforementioned , I you like it)

**Comment by randaz81** [comment]
I agree that the message chosen by @aerydna is too verbose. But users should be scared indeed, and educated to use this command as last resource. I often see yarpclean used to speed up yarpmanager refresh button. Even worse, I see people teaching to use this command, which resolves ZERO problems, and only breaks the network. Look at this message:
```
>yarp clean
# If the cleaning process has long delays, you may wish to use a timeout,
# specifying how long to wait (in seconds) for a test connection to a port:
#   yarp clean --timeout 5.0
Requesting list of ports from name server... got 2 ports
No timeout; to specify one, do "yarp clean --timeout NN.N"
```
Do you see any warning to prevent hasty users to request a yarp clean --timeout 0.01?(*) I think that adding one more line to the existing help is a good starting point. ROS node cleanup says:

`Warning: these might include alive and functioning nodes, e.g. in unstable networks.`

We can use a similar message.

(*) BTW: timeouts smaller than 0.1s should print an error message.

**Comment by pattacini** [comment]
I 100% agree on pointing out a warning message.

I don't agree on defining a threshold for the timeout that will pop up an error. Timeouts are anti-pattern by definition, hence there's no good threshold covering all conditions on all machines (especially in the single machine configuration).

Philosophically, I'd prefer much more to promote the _laissez-faire_ to risk and eventually learn by mistake than strict rules for quite (tiny) corner-case behaviors 😉

**Comment by barbalberto** [comment]
I do not agree. I love `yarp clean` command, I use it tons of times a day and it does helps cleaning up server port list from dead ones.
Saying that it **_only_** _breaks the network_ it's not true. It may does, if the network connection is badly compromise which, by the way, makes **all** commands unreliable.

If we want to add this message, please do a 1-line statement, like the ROS one, not that ugly big banner as originally in this PR.

**Comment by pattacini** [comment]
Didn't read the current proposal, actually. I tend to support a _sober_ and _concise_ warning message too.

**Comment by Nicogene** [comment]
This warning is already present when running `yarp clean` from `yarpmanager`. :wink: 
For me: warning :-1: documentation :+1:

**Comment by aerydna** [comment]
in my opinion is legit for a user to run a "clean" command without reading the documentation. From a clean command i **don't expect anything dangerous**. if the command was "yarpremove --deadport", maybe thinigs would be different but in a case like an anonymous "clean" the user should **notice at all cost** what is about to happen.

**Comment by aerydna** [comment]
@Nicogene the problem with yarpmanager is not for it to not having the warning... the problem is that if we add the user confirmation in the command line is not possible to launch it from yarp manager anymore.

**Comment by aerydna** [comment]
asking the user a confirmation is a **strong** choice. It is the best choice in my opinion because when the user should confirm something usually read carefully what is happening and ros chooses to do it. if we don't want to do it or if we can't we need a **strong** alternative to it

**Comment by barbalberto** [comment]
> in my opinion is legit for a user to run a "clean" command without reading the documentation.

Then run `rm -rf /` without reading the documentation (it's way more dangerous and it's not asking for confirmation).
I mean, the user is not required to know _how_ something is done by the framework, but at least he/she have to  know _what_ it is doing.
People not reading documentation or terminal messages is a user problem we cannot solve.

World is not idiot-proof, yarp cannot be either. People will make mistakes and learn from them.

Don't dare adding the requirement for confirmation from terminal. I don't care what ROS is doing, we are not ROS.

Btw, it has been like this for ages, why all this discussion now? I do appreciate new ideas, but I think here we are doign "much ado about nothing"

**Comment by Nicogene** [comment]
Documentation 👍 
Warning 👎 
User Confirmation 💀 
I know that `yarp clean` could be dangerous but in my opinion not enough to pollute the output and even less for add user confirmation. 
By the way this warning will be read by **nobody**.
Imagine to be on `iCubGenova01` setup with serveral modules running and something goes wrong. So I decide to run `yarp clean`.. the warning will not stop me to run it and it will be almost impossible to read for the huge number of requests for checking which ports are actually dead.

**Comment by aerydna** [comment]
rm -rf / don't have meaningful word in it. yarp clean, on the contrary sounds harmless. it's not about making it idiot-proof. it is about making it more user friendly. Following your example, windows (setting aside all his problem) doesn't let you erase all your data without confirmation. it is more user friendly. i mention ros because his case can directly be compared to our and because in some cases is more user friendly. **BUT** if we don't care to improve our "user friendlyness" cause we think is enough **no problem with it**. I only want to know it.

 let me only say that for what it was my experience (as a relatively fresh yarp user) we have to improve a lot.

**Comment by aerydna** [comment]
> the warning will not stop me to run it and it will be almost impossible to read for the huge number of requests for checking which ports are actually dead.

and that's why is a mega verbose banner! 

@Nicogene ok with reading the documentation when you use a software.. but there are some things that speaks for themselves or that should have to. when you launch a software with --help or --verbose you don't check the documentation to see effectively what the program will do. For an **anonymous general** clean functionality i'm expecting it to **clean only garbage** and i tend to consider it harmless. but maybe i'm a special case and the majority of the users outside will think differently

**Comment by aerydna** [comment]
> Btw, it has been like this for ages, why all this discussion now? I do appreciate new ideas, but I think here we are doign "much ado about nothing"

i didn't start the discussion.. i'm only answering to your questions/statements. 

but it's true that we are spending too much time into this. So i call myself out of this. I'm ok with your decision. no problem for me!  :-)

**Comment by barbalberto** [comment]
> Following your example, windows (setting aside all his problem) doesn't let you erase all your data without confirmation. it is more user friendly. 
 
Sorry, but that's not correct.
If you _permanently_ remove a file from a _GUI_ (file manager or whatever) Windows ask for a confirmation, like linux does and the yarpmanager does.

If you _permanently_ remove a file from a _command line_, windows **_does not_** ask for confirmation, neither linux, neither `yarp clean`

> we have to improve  "user friendlyness"

That's true, but IMHO this is not going in that direction

**Comment by aerydna** [comment]
@barbalberto  i don't know which version of windows are you using but here i am in windows and calling del * in the user folder pops out a confirmation question

**Comment by drdanz** [comment]
Ok, this is going too far, I'll let you all have a big brawl, the winner will survive, so he will get to implement the policy that I decide 🤣.

For reference, my personal opinion is that `yarp clean` (and any other command doing something "dangerous") should request a user confirmation, (in this case, yes enabled by default), and have an option `-y` to bypass this confirmation for using it in scripts and similar.

This means that instead of writing `yarp clean<enter>` the user can either write `yarp clean<enter><enter>` or `yarp clean -y<enter>`. An extra `<enter>` does not seem a big issue to me. Also people is used to adding `--timeout 0.1`, so adding an extra `-y` should not be a big deal.


User will not read the documentation (mostly because none of the following works `yarp clean --help`, `yarp clean -h` `yarp clean help` `yarp help clean` `man yarp` `man yarp clean` `yarp clean \?`), having only an online documentation for a command line tool is very close to not having any documentation.


Also it's weird (in my opinion) that the client tries to connect to the ports and sends unregister commands to the server, i think this might be prone to race conditions (i.e. the client tries to connect a port that is being registered on the server and therefore not reachable. The client sends the unregister command, but meanwhile the port is opened, so now you might have an open port that is not registered on the server). IMHO, the whole clean task should be moved to the server that should not allow register/unregister commands during the execution of the task.

**Pull request closed without merging**